### DISCUSSION TITLE
Assisting

### DISCUSSION THREAD

**Discussion by larsnaesbye** [nav-solutions/rinex#5] [General]
Hi, I might want to contribute to the development of `rinex`, as my day to day work is in need of better tools than tecq, Anubis, gfzrnx etc. I have experience with RINEX, and though fairly new to Rust, I'm already fascinated by the tooling and community.

 Are there some smaller tasks you'd like to be worked on first, like contributing test data, hardening or fixing the epochs?

**Comment by gwbres** [comment]
Hello @larsnaesbye, welcome onboard!

At the moment it is fair to say that the parser is the most stable part of the project. 
Navigation, Observation, Crinex (compressed Obs) and Meteo records are very stable (up to V3). I just marked Antex and Clocks as stable in the last few days. You can refer to the entire `/test_resources` database, which is used for automated test purposes (for information, there are about 40 tests run automatically at the moment, mostly about the parser).

The `rinex` directory is a rust library, which everyone can include and take advantage of, the official documentation can be found [here](https://docs.rs/rinex/0.4.0/rinex/). It is shipped with a couple of examples, you can browse their codes in `rinex/examples` directory, I should get back to enhance them and provide more examples (clocks, antex..). 
For example, you can run the observation example with:

```shell
cargo run --example observation
```

I am most focused on the parser (file reader) at the moment as opposed to 
* the command line tools: which you are probably more interested in
* and the file writer (file production) which you might be interested in

I am currently concluding support for `IonosphereMaps` and `Troposphere` data, then I will get back to the command line tool.

Regarding the command line tools, `rinex-cli` is an application, written in rust. That means it builds easily with "cargo" but it is supposed to be run in a shell. It uses the main library (but hidden to the user), so the user can easily take advantage of it. The command line application is somewhat capable of analyzing OBS/NAV/MET. It is not capable of plotting at the moment, I will try to provide that in a near future. There is a lot that can still be achieved with this tool. If you want to give it a try, you can do something like this:

```shell
cd rinex-cli
cargo run -- --help 
# print available obscodes in this RINEX
cargo run -- -fp ../test_resources/OBS/V3/DUTH0630.22O --obscodes --pretty
# extract data for Sv "G01" but only for epochs with an `ok` flag attached to them:
cargo run -- -fp ../test_resources/OBS/V3/DUTH0630.22O -e --sv G01 --pretty
```

As for the `teqc`-like operations, which you are interested in, there are three things we can say:

* the action takes place at the Rinex structure level, see [split_record_at_epoch()](https://docs.rs/rinex/0.4.0/rinex/struct.Rinex.html#method.split_record_at_epoch) and neighboring methods,
like `merge()`. 
* they are not too complex, but very verbose.. because the core differs for every record type (hidden to user).
I barely spent any time on these methods at the moment. I would say, the best thing to do is to focus on the `merge()` ops, and test it properly in `rinex/tests/merge.rs`. I am also looking forward to providing a `down_sample()` method.
* once we stabilize such methods, it can easily be exposed in the command line tool to create a probably similar (and possibly more advanced) `teqc` operation

Overall, and I will conclude with this:
* I started using RINEX this year and I still have doubts / questions about things. 
I also struggle to find any test data, I don't have a single V4 for instance.
If we can discuss things in the future, just that would be already valuable help. My job involves timing and related measurements (hence the ClockData support), I am looking forward to providing proper Time System conversions and related calculations
* the record manipulation is the most complex thing in this project. First of all because the record type depends on the type of file, so they are all different. Then, because it involves iterator. Iteration is very efficient and powerful, it provides generic filter and sorting methods (see the example codes), but it is also very verbose and hard to grasp in my opinion. 
* for example, OBS/NAV/MET are all iterated on an `epoch` basis, a `merge` operation will obviously involve epoch iterations
* So far I have used 3 structures to declare Data Arrays:
  * `Vec` : standard Vector indexed by an integer, they are not sorted! order is unpredictable
  * `Hash`: Vector but indexed by an object (which must implement `==` comparison). I used `Space vehicule` indexing a lot, which sorts data efficiently
  * `BTreeMap` : same as `Hashmap` but data is ordered! (must implement `Ord`). I use BTreeMaps to have the record chronologically sorted on an epoch basis
 
Now that is already a lot of information. Let me know if you want to know more.

**Comment by larsnaesbye** [comment]
Hello @gwbres,

Thank you for the very descriptive welcome! There is absolutely everything I would like to know in there for now.

You are right, the file production and command-line tools are what I ultimately would like to use for daily operations. I oversee a system that produces RINEX files from raw receiver data and distributes them freely for public consumption.

In short, the tasks are:

* Convert raw binary from receiver to RINEX - obviously not in scope for this project
* Pack and unpack gzip
* Compress and decompress Hatanaka
* Rewrite RINEX headers
* Splice 1-hour files into 24-hour files
* Decimate 1-sec files into 30-sec files
* Gap analysis
* QC on the 30-sec files

For this we use a combination of gfzrnx, bnc, Anubis and CRX2RNX/RNX2CRX (the one by Hatanaka). If these tasks could in the future be done using the same framework, it would make everything easier and more transparent. 

I think I can pretty easily get some RINEX v4 files, since the RAW to RINEX software supports it. Let me get back to you on that.

After that I might look into the `rinex-cli`, if that's OK with you.

**Comment by gwbres** [comment]
Hello @larsnaesbye ,

>I oversee a system that produces RINEX files from raw receiver data and distributes them freely for public consumption

I initiated a "ublox-rinex" application, it is also contained in this folder, and it serves that role as long as you intend to use "ublox" receivers. It basically combines the "ublox" crate (very powerful! I use it at work on a daily basis), and the "rinex" crate (which already knows the most common RINEXes), that could be of interest for you. I barely spent any time on this application. We need to conclude the `Rinex::to_file()` method to make this application happen. Which is not too painful and should not take long. After that, we can combine the best of both worlds. If you intend to use different receivers, obviously that will not help.

>Convert raw binary from receiver to RINEX - obviously not in scope for this project

The `rinex-ublox` app does that from a ublox receiver. The official `ublox` crate lacks some Ublox frames, therefore I have the application point to my fork (timing, and other ubx frames)

>Pack and unpack gzip

I just added the hability to parse `.gz` files directly. I am improving the `BufferedReader` (most efficient way to browse the file) in the sense it is already capable to uncompress .gzip in a hidden fashion and soon Hatanaka too (makes things much easier).
If we start tackling the `to_file()` side of things, we would need the mirror operations, `BufferedWritter` for efficient file production and smooth operations.

>Compress and decompress Hatanaka

Supported natively, will be managed by the `BufferedReader` directly in my next commits.
Note I have not completed the Hatanaka compression (mirror operation) in the Hatanaka core structures! needs to be finalized to have an efficient `to_file()`

 >Rewrite RINEX headers
 
[Rinex::to_file()](https://github.com/gwbres/rinex/blob/main/rinex/src/lib.rs#L948) is partial, it needs to be improved, support a buffered stream efficient approach, support all known attributes and record types..
 
 >Splice 1-hour files into 24-hour files
 
I have not initiate the Rinex::splice operation, because I do not know what splicing means

>Decimate 1-sec files into 30-sec files

Decimate/down sampling is under development but is closer to stable than all other topics.
[See Rinex::resample(std::time::Duration)](https://github.com/gwbres/rinex/blob/main/rinex/src/lib.rs#L789) to rework a sampling interval. Or [Rinex::Decimate(N)](https://github.com/gwbres/rinex/blob/main/rinex/src/lib.rs#L897) to reduce the epoch rate by N

>Gap analysis

[See Rinex::dead_times()](https://github.com/gwbres/rinex/blob/main/rinex/src/lib.rs#L342) for gap analysis, also [Rinex::anomalies()](https://github.com/gwbres/rinex/blob/main/rinex/src/lib.rs#L372) for epoch flags analysis (flags only truly relevant in OBS files)

**Comment by larsnaesbye** [comment]
> I initiated a "ublox-rinex" application, it is also contained in this folder, and it serves that role as long as you intend to use "ublox" receivers. It basically combines the "ublox" crate (very powerful! I use it at work on a daily basis), and the "rinex" crate (which already knows the most common RINEXes), that could be of interest for you. I barely spent any time on this application. We need to conclude the `Rinex::to_file()` method to make this application happen. Which is not too painful and should not take long. After that, we can combine the best of both worlds. If you intend to use different receivers, obviously that will not help.

Yeah, we use and have used other types of receiver (Leica, Trimble, Septentrio...), not u-blox.

> I have not initiate the Rinex::splice operation, because I do not know what splicing means

I think splice and merge is the same thing - combining two files from the same receiver to get one continuous. A bit like they describe on the tecq mailing list here:  https://postal.unavco.org/pipermail/teqc/2016/002144.html 

I have submitted a pull request with a RINEX 4 file (Hatanaka 3 compressed). I hope it can be of use for testing :-)

**Comment by gwbres** [comment]
Hello @larsnaesbye 

I validated some high level record operations, I started integrating them into `rinx-cli`, which will reduce the code quantity and make things much smoother / efficient. In `rinex-cli`, by using the `mut()` declination of filters, we can stack them together and create interesting operations.

Here is my notation:
:heavy_check_mark:  coded, tested, integrated to autotests
:sparkler: should work

* `decimate_by_interval(duration)`   :heavy_check_mark: 
At the moment, it is not a feasible operation if the header section did not come with an *INTERVAL* field.
Which is a bit silly.. A method to determine the most plausible epoch interval (average, histogram peak..) should exist

* `decimate_by_interval_mut(duration)`   :heavy_check_mark: 
same thing, but decimates in place

* `decimate_by_ratio(integer)`   :heavy_check_mark: 
same thing but works as a modulo operation, not from epoch intervals

* `decimate_by_ratio_mut(integer)`   :heavy_check_mark: 
same thing but decimates in place

* `lli_filter (mask)` : :sparkler:  
applies AND mask on LLI data in this observation rinex.
Has no effect on Non Observation Data.

* `lli_filter_mut(mask)` : :sparkler:
same thing, but works in place 

* lock_loss_epochs: :sparkler:
returns list of epochs where lock was declared lost. 
* lock_loss_filter: :sparkler:
cleans up epochs (in place) returned in previous method

* `epoch_ok_filter_mut()` :sparkler: but i'm confident
* `epoch_ok_filter()` :sparkler: but i'm confident

* `data_gap` :sparkler:
returns list of data gaps, meaning epochs where |e(k)-e(k-1)| > interval

* `constellation_filter_mut`  :sparkler:
* `space_vehicule_filter_mut`  :sparkler:
* `average_epoch_duration`   :sparkler:

**Comment by larsnaesbye** [comment]
This is really good. I think I will soon try to run a process flow and see if I can do the same with `rinex-cli` as I use 4-5 tools for today.

**Comment by gwbres** [comment]
@larsnaesbye 

I added a note to the documentation on [how to browse the different record types](https://github.com/gwbres/rinex/tree/main/rinex#rinex-record)

Here are a few commands I can execute on my side, they seem to produce the expected results.
File generation is about perfect for meteo files, about 90% concluded for OBS, and just started for NAV.
I'll inquire the next PR tomorrow or in the next few days.
Once rinex-cli is stabilized, I will unlock an option to plot the data (GUI), not just print it, for a better user experience
 
```bash
# --decim-ratio : reduce information quantity
rinex-cli -- -f test_resources/OBS/V2/aopr0010.17o -o --decim-ratio 4 --pretty

# --decim-interval : provide a duration and reduce information efficiently
cargo run -- -f ../test_resources/OBS/V2/zegv0010.21o \ 
          --ssi 8 # excelent signal quality \
                 --decim-interval 00:01:00 # new option \
                         --pretty

# focus on GPS vehicules
cargo run -- -f ../test_resources/OBS/V2/zegv0010.21o \
             --constellation GPS \
                       --decim-interval 00:02:00 --pretty
                     
                     
# focus on G07
cargo run -- -f ../test_resources/OBS/V2/zegv0010.21o --sv G07 --decim-interval 00:02:00  --pretty

# split at epoch
# split operation, with an epoch specified, 
# we disregard whether this file is merged or not, specified epoch must exist
cargo run -- -f ../test_resources/OBS/V2/zegv0010.21o --decim-interval 00:01:00 --split "2021-01-01 00:09:00"
# note the "quotes ‘’, needed for the whitespaces

# this produces the couple (split1.txt, split2.txt). 
# There you see what I just pushed, with the first few to_file() methods made avaialble.
# There is no mean to properly control the names to be generated at the moment

**Comment by larsnaesbye** [comment]
Looking good. I noticed that the v4 NAV file does make the tests fail. I might look into this too.

**Comment by gwbres** [comment]
Hello

@larsnaesbye 
you might want to take a look at the methods I pushed today,

* [space_vehicule_clocks_offset()](https://github.com/gwbres/rinex/blob/main/rinex/src/lib.rs#L925)
* [space_vehicule_clocks_drift()](https://github.com/gwbres/rinex/blob/main/rinex/src/lib.rs#L985)
* [pseudo ranges to geometric distances](https://github.com/gwbres/rinex/blob/main/rinex/src/lib.rs#L1434) 
I'm not sure this one has an application ? 
we can obtain the space vehicules clock offsets from NAV frames, 
we might have the local clock offset on an epoch basis for this OBS,
we can convert PR to real distance
* [iono_free_pseudo_ranges()](https://github.com/gwbres/rinex/blob/main/rinex/src/lib.rs#L1250)
if we have observations from two separate carrier, on an epoch and vehicule basis
* [iono_free_carrier_phases()](https://github.com/gwbres/rinex/blob/main/rinex/src/lib.rs#L1334)
if we have observations from two separate carrier, on an epoch and vehicule basis

**Reply by larsnaesbye** [reply]
Thanks. Will look into it tomorrow :)

**Reply by gwbres** [reply]
Hello @larsnaesbye 

As I am working in the nav_v4 branch in advance, prior merging your PR, I just noticed, and fixed significant issues regarding NAV support. Unfortunately you need to wait for me to conclude all of this prior being able to take advantaged of all the fixes. 

Mainly:
* the spare fields were not described properly, which led to the Orbits Data to be uncorrectly interprated ("shift" effect in description keys)
* BeiDou and GAL descriptors had some mistakes
* the JSON library that interprated the NAV dictionnary, automatically sorted the keys by alphabetical order, which completely messed up the database which no longer correspond to the specifications: data badly interprated

Due to NAV V4 being so different from the rest, the browing will be heavily impacted.
I will need a few more days to stabilize all of it, in the end, nav_v4 and your PR get merged

**Reply by larsnaesbye** [reply]
No hurry for me. I’m very impressed already, and our use case for RINEX v4 is limited at the moment :)

**Reply by larsnaesbye** [reply]
Hello @gwbres

Before I go on 3 weeks vacation on Wednesday, are there other types of sample data you would like me to see if I can produce?

**Reply by gwbres** [reply]
Hello @larsnaesbye 

thank you for asking that's very kind of you. 
I think we're reaching a steady state, we span most kinds and revisions.
Some IONEX (ionosphere maps) if you happen to have such data, could of interest because it is rare to find.
I found some meaningful clock data on my side, I will commit about this topic in the next few days.
After that I will mostly focus on the application, I will focus on teqc-like features. Right now I am working on the ascii plot and tiny report. Do you foresee any command line / scripting / plotting feature that could really be helpful or convenient to have? 

Enjoy your vacation time ! :sun_with_face:

**Reply by gwbres** [reply]
Hello @larsnaesbye ,

the next release of the command line tool will have the ability to perform single and double difference RINEX post processing.
The functions are [coded here](https://github.com/gwbres/rinex/blob/a4d4f6b97956bb7b65bf5935b5efdf22efbb7b6c/rinex/src/lib.rs#L2851).
The single difference (rnx = rnx_a - rnx_b) is easy to perform and should work as expected (cancels out ionospheric biases).
For the double difference, we designate a reference vehicule for each constellation for each epoch, and compute the phase difference between other vehicules and this reference. I am not sure how people select the reference vehicule and if that is important.
See [double_diff()](https://github.com/gwbres/rinex/blob/a4d4f6b97956bb7b65bf5935b5efdf22efbb7b6c/rinex/src/lib.rs#L2931)
and [here I thought I could let the user pick the reference(s) vehicule(s)](https://github.com/gwbres/rinex/blob/a4d4f6b97956bb7b65bf5935b5efdf22efbb7b6c/rinex/src/lib.rs#L3002)

**Reply by larsnaesbye** [reply]
Very nice! Looking forward to test it when I get back to work :)

**Reply by larsnaesbye** [reply]
Hello again @gwbres ,

I'm back from vacation, and it is real nice to read the thoughts on implementing much of `tecq` functionality. Many around the world still hang on to it because of its features, even though it is discontinued, it only supports RINEX v2, and the source is not open for modification.

If this toolkit will emulate a big part of that functionality, I see both a reason to adopt it myself and to promote it to partners.

One thing I thought of doing myself is to write an importer/converter from the proprietary SBF (Septentrio) format, which I use myself. If I get it done and it fits in with the scope, I'd be happy to eventually donate it to this project. I have no time perspective for it though.

**Reply by gwbres** [reply]
Hello @larsnaesbye 

I hope you enjoyed your vacations

I have some doubts about what should be done in the double difference process, when it comes to selecting the "reference" vehicule. 
At the moment, I arbitrarily define the first vehicule encountered (for a given constellation) as the reference. Should this vehicule remain the reference accross all epochs in the file ? or should we define a reference vehicule (that might differ) for each epoch? Because it is highly possible that the selected vehicule might be missing in future epochs - therefore, no feasible differentiation when the "reference vehicule" gets out of sight.

Also, is the reference vehicule important? currently I define the first encountered, but in real application, does the end user want to do that himself ?

Finally I am not familiar with the final, so called RTK, post processing, which involves related Ephemeris to be attached (vehicule position vectors). If you are familiar with such operation, it would be very helpful if you could elaborate a little on what needs to be done,
thank you

**Reply by larsnaesbye** [reply]
Sure. I went straight from vacation to a conference (FOSS4G) so I have not had any office hours and time to test properly yet. I will try to prioritize a test run next week.

**Reply by larsnaesbye** [reply]
Hello @gwbres 

I spoke to my knowledgeable colleagues and they say that often the satellite that is highest in the sky (closest to zenith) is used as reference, and that this can be inferred from the ephemeris as you mention.

**Reply by gwbres** [reply]
hello @larsnaesbye 
thank you for getting back on this. That makes sense, I suppose signal quality is better.
Btw we already have high level methods to [retrieve](https://github.com/gwbres/rinex/blob/main/rinex/src/lib.rs#L1752) and [filter](https://github.com/gwbres/rinex/blob/main/rinex/src/lib.rs#L1792) on elevation angles.
Although I can't remember whether I expose them to the `cli` in current version. 
I have been kept busy on other things lately, I will get back to it in the next few days.

**Reply by larsnaesbye** [reply]
Hello again @gwbres 
Happy to help!
Over the next couple of days I will test the CLI with the common operations I currently use other tools for.

**Reply by larsnaesbye** [reply]
By the way, if one wanted to rewrite or set RINEX headers like Observer/Agency, would it be within scope to add this functionality?

**Reply by gwbres** [reply]
Hello @larsnaesbye,

> Over the next couple of days I will test the CLI with the common operations I currently use other tools for

No hurries, this will be valuable input

> if one wanted to rewrite or set RINEX headers like Observer/Agency, would it be within scope to add this functionality?

You currently have two options to customize / manage header fields.
Either by customizing specific fields:

```rust
let mut rinex = Rinex::from_file("somefile.rnx")
      .unwrap();
let mut header = rinex.header;
header = header
        .with_general_infos("Program", "RunBy", "Agency");
header.agency =  "MyAgency".to_string(); // direct access is permitted
         // because currently all attributes are declared `public`,
     // I did so because I thought it would be tedious to provide one method for each field
```

Or by building a `Header` of your own

```rust
let rinex = Rinex::from_file("somefile.rnx")
      .unwrap();
let customized = Rinex::Header::basic_nav(); // useful?
let customized = Rinex::Header::basic_obs(); // useful?
let customized = customized
       .with_rcvr(Rcvr {
            model: String::from("Model"),
            sn: String::from("SN#"),
            firmware: String::from("Version1"),
       });
let rinex = rinex.with_header(customized);
rinex
      .to_file("output.rnx")   // hopefully I will conclude this one day... -_-
      .unwrap(); 

let customized = customized
     .with_crnx(Crinex {
          version: String::from("CRNX3"),
          prog: String::from("custom"),
          date: Utc::now().date_naive(),
    });
let rinex = rinex.with_header(customized);
rinex
    .to_file("output.crnx") // and the magic happens
    .unwrap();
```

**Reply by larsnaesbye** [reply]
Hey @gwbres 
> You currently have two options to customize / manage header fields. Either by customizing specific fields:
> 
...
> 
> Or by building a `Header` of your own
> 
Sure, that makes perfect sense. I was more thinking of setting headers via JSON as an option in `rinex-cli`, e.g.:

```bash
rinex-cli --filepath AMEL00NLD_R_20210010000_01D_MN.rnx --output updated/AMEL00NLD_R_20210010000_01D_MN.rnx --specify-headers {"OBSERVER / AGENCY" : "UNOOSA", "COMMENT" : "Not for redistribution"}
```

This could be something I could work on unless you think it is outside the scope for `rinex-cli`.

**Reply by gwbres** [reply]
>unless you think it is outside the scope

totally not, I think it's a good idea and becomes needed quickly in real production context like yours.
It's just I had not thought about it 

>I was more thinking of setting headers via JSON as an option

It's actually a good idea and probably more efficient than specifying fields one at a time.
Maybe keeping the option to also grab it from an external JSON file will be convenient 

>This could be something I could work on

absolutely. The structure `Serialization` and `Deserialization` is where to go.
At the moment, the `Header` structure can only be serialized, that means it can be dumped into a JSON structure in one line, something like this

```rust
serde_json::to_string(my_header)
```

For the whole structure to be ser/deserialisable, all its inner members must be ser/deserialisable. 
I did not implement the `Deserialization` trait for the header, because it was not required until now.
For special structures that apparently do not have a "serdes" trait implementation, I coded everything in `rinex/src/formatter.rs` to reduce the code quantity of other files like `header.rs`. Encapsulating the `serialize` `deserialize` functions inside modules like I did there, makes it very easy to use when auto-implementing the Trait (see what happens in `pub struct Header`).

After a quick glance, the missing (unimplemented) Deser ops are:
- `formatter.rs::point3d::serialize` mirror operation (you can find the appropriate `deserialize` function signature online)
- datetime structure, same location
-  unfortunately all RINEX specific fields seem to lack the `Deser` trait, structures like `src/observation::HeaderFields`

**Reply by gwbres** [reply]
Hello @larsnaesbye,
I was thinking about how to solve the topic you are interested in.

There's the problem of dealing with the amount of data fields you are providing (from the command line, or from the file).
If you make the structure deserializable, that means a valid deserialization is always a complete `Header`, which is huge, very impractical and almost impossible to achieve.  For example let's say you are producing OBS RINEX, you absolutely do not care about METEO related fields. Most of the time the user will want to customize very general fields (like Observer, Agency) and does not care at all about other fields of the header section.

There might be something smart to do with `skip`  option of the derived `deserialization`. Maybe declaring the most important / most common fields as impossible to skip, and then make all other members as possibly skipped, which will reduce the descriptor to pass to the command line, and gives the user enough flexibility to possibly describe the structure in its entirety.
This would not impact the library at all.

Another option is to rewrite the `Header` structure and make all members (except the mandatory ones) as `Option<>`, I am no Serdes expert, but I think this makes the `deserialization` naturally handle empty description / missing fields. On the other hand, this would require some portions of the `Parser` to be rewritten

**Reply by gwbres** [reply]
Hello

@larsnaesbye 

I just unlocked `Header` deserialization.   
All Header fields that are not declared as "Optionnal" are required in the Header JSON descriptor.  
You know better than I do, I think we could use more optionnal fields. 


❗  `--with-serdes` feature is renamed `serdes`: apparently this is the standard and it is a special case  ❗  

As far as the CLI goes, I introduced a --header-json argument: 
* if Deserialization works from its content, that means the user passed a valid description directly
* otherwise we try to interprate its content as a File path, and we try a Deserialization from its content
* if both fail, that means the user is not providing expected content

**Reply by larsnaesbye** [reply]
That's really cool that you implemented this! I am sure many organisations have a need to set the headers to indicate where data comes from.

**Reply by gwbres** [reply]
Hello @larsnaesbye 

by curiosity, would enhanced or improved performances, as opposed to supposedly less performant solutions (like interpreted python for instance), be something interesting to you? or at least make this library more valuable than other solutions

at the moment, the main interest I see in this library is the ability to process any kind of rinex (well.. almost, we'll get there), and span all revisions. I guess few other tools have this ability. The python lib I know is limited to OBS and NAV (not even sure) and probably has some revision limitations. Also I'm pretty sure very few support V4 like we do. But people did not take the V4 gap yet, I assume because of the previous point.

I hear often that Rust code is supposedly "more performant". It's compiled so we should out perform python libraries. But I also hear that it can achieve blazing performances. I know some portions of this lib can seriously be improved, i'm thinking about the OBS and NAV reading part. The compression part is quite optimized at the moment I would say. Anyway, this would come in a second phase where all major topics were completed first. I'll probably inquire some means to evaluate parsing/compressing performances at some point.

see you

**Reply by larsnaesbye** [reply]
Hmmm, I wonder which tools would be good for finetuning Rust performance.

**Reply by gwbres** [reply]
> Hmmm, I wonder which tools would be good for finetuning Rust performance.

I looked into this a little bit, there is a benchmark infrastructure that can be added on top of builtin tests.
It is invoked in similar fashion, and has lots of objects and features to evaluate performances.
The only thing is, benchmarks can only run against the unstable rust compiler to this day, so it has to be isolated from the rest of the library.
BTW, in case you don't use it: I highly recommend using `rustup` to manage your Rust environments - its like a python manager. 
I personally cross-compile rust code a lot, and it is very easy to do so with this tool.
I just ran a quick test of this whole benchmark thing, the test parsed the gzip compressed NAV4 you recently provided, you gets basic results like these:

```shell
rustup default nightly # set unstable rustc as default
rustup run nightly cargo bench # invoke "cargo bench" by temporarily switching to unstable compiler (very convenient)
bench:  19,910,022 ns/iter (+/- 835,513)  # 20 loops
bench: 103,932,756 ns/iter (+/- 3,091,624) # 100 loops, 
```

I'll probably add a benchmark infrastructure because it won't require much effort.
Results don't seem very meaningful to me, but it will serve as a comparison tool if we track their evolution

**Reply by larsnaesbye** [reply]
I mainly used rustup for initial setup, but good to know it’s like `conda` for Python.

**Comment by gwbres** [comment]
Hello @larsnaesbye 

This is a little bit off topic, but do you have any recomendations in terms of GNSS receiver settings, in order to maximize Timing/Solution quality? Right now I'm investigating Elevation mask and so called "static hold" parametrization. I only have access to Ublox (8) receivers, Unfortunately we don't have access to high end Ublox (9 model) or Septentrio receivers.

**Comment by gwbres** [comment]
Hello @larsnaesbye

can you confirm that we also (possibly) want to compute 1D or even 2D differentiations, or raw pseudo ranges, and not only raw phases ?
currently I discarded anything but raw phases I think, I will improve that.

If you request a 1D or even worse a 2D diff, with the realistic context you provided last summer, the operation is very very slow on my side. I can only produce a result when focusing on 1 or 2 vehicules. But we have a good progression margin, on both record parsing and crinex decompression, so i'm expecting this to improve in a near future.

PS: Observation file generation about to be concluded :)

**Reply by larsnaesbye** [reply]
I am not quite sure about the need for 2D differentiations. I will ask my colleagues next week when they come home from autumn holidays :)

**Reply by gwbres** [reply]
> I am not quite sure about the need for 2D differentiations. I will ask my colleagues next week when they come home from autumn holidays :)

Holidays are starting here, enjoy your vacation time :+1: 
For information, a friend of mine pointed out 
[this american document](https://www.gps.gov/technical/icwg/IS-GPS-200N.pdf), 
which seems to be the reference document for everyone working on RTK differential positionning for instance. I will study it, it looks like it contains all answers to my questions

**Comment by gwbres** [comment]
Observation RINEX encoding and formatting is now fully validated :+1:.
I'm providing a file generation interface in the command line as of now, I'm in the process of documenting this section.

Validating V2 has honestly been a real pain.
And it looks like our parser faces a weird issue, like it's working 99.99 % of the time.
When parsing `test_resources/others/rovn` (temporarily moved out of pool), one epoch comprises 25 vehicules, which requires a 3 line descriptor. We know we have 25 vehicules and know we must grab 2 extra lines. But we wind up grabbing 3 lines, which destroys the 1st observations. Mistery... :space_invader:
